src/scratchOrgCreateCommand.ts

```diff
@@ -0,0 +1,271 @@
+    const definitionJson = this.options.flags.definitionjson ? JSON.parse(this.options.flags.definitionjson) : {};
+    const orgConfigInput = { ...definitionJson, ...(this.options.varargs ?? {}) };
+
+    let scratchOrgInfoPayload = orgConfigInput;
+
+    // the -f option
+    if (this.options.flags.definitionfile) {
+      try {
+        const fileData = await fs.readFile(this.options.flags.definitionfile, 'utf8');
+        const defFileContents = parseJson(fileData) as Record<string, unknown>;
+        // definitionjson and varargs override file input
+        scratchOrgInfoPayload = { ...defFileContents, ...orgConfigInput };
+      } catch (err) {
+        const error = err as Error;
+        if (error.name === 'SyntaxError') {
+          throw new SfdxError(`An error occurred parsing ${this.options.flags.definitionfile}`);
+        }
+      }
+    }
+
+    // scratchOrgInfoPayload must be heads down camelcase.
+    const upperCaseKey = sfdc.findUpperCaseKeys(scratchOrgInfoPayload);
+    if (upperCaseKey) {
+      throw new SfdxError('InvalidJsonCasing', upperCaseKey);
+    }
+
+    // Now run the fully resolved user input against the validator
+    Object.keys(scratchOrgInfoPayload).forEach((key) => {
+      optionsValidator(key, scratchOrgInfoPayload);
+    });
+
+    // the -i option
+    if (this.options.flags.clientid) {
+      scratchOrgInfoPayload.connectedAppConsumerKey = this.options.flags.clientid;
+    }
+
+    // the -d option
+    scratchOrgInfoPayload.durationDays = this.options.flags.durationdays;
+
+    // Ignore ancestor ids only when 'nonamespace' or 'noancestors' options are specified
+    const ignoreAncestorIds = this.options.flags.nonamespace || this.options.flags.noancestors || false;
+
+    // Throw warnings for deprecated scratch org features.
+    const scratchOrgFeatureDeprecation = new ScratchOrgFeatureDeprecation();
+    scratchOrgFeatureDeprecation.getFeatureWarnings(scratchOrgInfoPayload.features).forEach((warning) => {
+      warnings.push(warning);
+    });
+
+    const scratchOrgInfo = await generateScratchOrgInfo({
+      hubOrg: this.options.hubOrg,
+      scratchOrgInfoPayload,
+      nonamespace: this.options.flags.nonamespace,
+      ignoreAncestorIds,
+    });
+
+    return {
+      scratchOrgInfo,
+      warnings,
+    };
+  }
+
+  private async updateRevisionCounterToZero() {
+    const conn = this.scratchOrg.getConnection();
+    const queryResult = await conn.tooling.sobject('SourceMember').find({ RevisionCounter: { $gt: 0 } }, ['Id']);
+    try {
+      await conn.tooling
+        .sobject('SourceMember')
+        .update(queryResult.map((record) => ({ Id: record.Id, RevisionCounter: 0 })));
+    } catch (err) {
+      const message = messages.getMessage('SourceStatusResetFailure', [
+        this.scratchOrg.getOrgId(),
+        this.scratchOrg.getUsername(),
+      ]);
+      throw new SfdxError(message, 'SourceStatusResetFailure');
+    }
+  }
+}
```

Comment:
We shouldn't do this as part of scratch org creation.  This is for source tracking and the caller may never want/use it.  Source tracking initialization should handle this.